Wasm threading fixes

PgBiel opened the pull request:
Currently implemented:

 - Adds custom wasm binary name support proposed in https://github.com/godot-rust/gdext/issues/438#issuecomment-2735355627, sample usage:

      ```rs
      struct MyExtension;
    
      #[cfg(feature = "nothreads")]
      #[gdextension]
      unsafe impl ExtensionLibrary for MyExtension {}
      
      #[cfg(not(feature = "nothreads"))]
      #[gdextension(wasm_binary = "mycrate.threads.wasm")]
      unsafe impl ExtensionLibrary for MyExtension {}
      ```
- Fixes `wasm_nothreads` compilation by not using compiled out function in `async_runtime.rs` (regression from https://github.com/godot-rust/gdext/pull/1043)
- Fixes `wasm_nothreads` runtime errors on init due to thread locals being used for panic handling (regression from https://github.com/godot-rust/gdext/pull/1037)

 ## TODO

Working on these, but feel free to review or merge what's already there.

- [x] ~~Potentially fix the rest of `async_runtime` to e.g. not use `std::thread::current` on nothreads builds (might need to be in a separate PR~~ Let's do that in a separate PR (edit: using that function is actually fine, let's keep an eye on it in case anything else could be a problem)
- [x] Automatically detect `.threads.wasm` when it still uses the crate name to avoid having to use the option introduced above outside atypical cases

GodotRust commented:
API docs are being generated and will be shortly available at: https://godot-rust.github.io/docs/gdext/pr-1093

GodotRust commented:
API docs are being generated and will be shortly available at: https://godot-rust.github.io/docs/gdext/pr-1093

Bromeon reviewed (commented):
Thanks a lot!

PgBiel commented on godot-macros/src/gdextension.rs:
```diff
@@ -48,6 +49,13 @@ pub fn attribute_gdextension(item: venial::Item) -> ParseResult<TokenStream> {
         .unwrap_or_else(|| ident("gdext_rust_init"));
 
+    let wasm_binary = match wasm_binary {
+        Some(wasm_binary) => {
+            quote! { format!("'{}'", (&#wasm_binary as &str).replace("\\", "\\\\").replace("'", "\\'")) }
+        }
```
Well, I'm using `handle_expr` here so `wasm_binary` is a black box. That's also why I'm using the cast, to ensure the right type is passed. In a way it is nice in that it allows passing a `const`, but I suppose we could try to force a literal for the time being? I couldn't find a way to force a string literal specifically, maybe we can just check if it starts and ends with quotes

Bromeon replied:
Yeah, but even then you could do
```rs
// Expect string-compatible expression.
let wasm_binary_str: &str = &#wasm_binary;
```
At least this avoids the weird `as` cast, which I've never seen used in this context...

PgBiel commented:
News:
- I've implemented a built-in check for `.threads.wasm` so the book instructions can be more easily followed. The `wasm_binary` option is now only used for non-standard names.
- I've added the requested docs, not sure if the example is too ideal given this, let me know if it should be changed or if it's ok for now

Bromeon commented on godot-macros/src/gdextension.rs:
```diff
@@ -48,6 +49,19 @@ pub fn attribute_gdextension(item: venial::Item) -> ParseResult<TokenStream> {
 
+    let wasm_binary = match wasm_binary {
+        Some(wasm_binary) => {
+            quote! {
+                {
+                    // Ensure a string-like expression is passed
+                    let wasm_binary: &str = &#wasm_binary;
+                    format!("'{}'", wasm_binary.replace("\\", "\\\\").replace("'", "\\'"))
+                }
+            }
```
Adding a comment, because knowing myself, I'll be tempted to extract this in the future :grin: 
```suggestion
            // Replacement is not moved out of quote! to allow non-literal expressions (e.g. constants).
            quote! {
                {
                    // Ensure a string-like expression is passed.
                    let wasm_binary: &str = &#wasm_binary;
                    format!("'{}'", wasm_binary.replace("\\", "\\\\").replace("'", "\\'"))
                }
            }
```

Bromeon replied:
Furthermore, should we validate that this ends with `.wasm`?

PgBiel replied:
> Furthermore, should we validate that this ends with `.wasm`?

Yep, added it.

Bromeon commented on godot-core/src/init/mod.rs:
````diff
@@ -244,6 +244,38 @@ fn gdext_on_level_deinit(level: InitLevel) {
+/// `-with-threads.wasm`. For this, you can have a `"nothreads"` feature which, when absent, should produce a suffixed binary, which
+/// can be informed to gdext as follows:
+///
+/// ```no_run
+/// # use godot::init::*;
+/// struct MyExtension;
+///
+/// // Tell gdext we add a custom suffix to the binary with thread support
+/// // Please note that this is not needed if "mycrate.threads.wasm" is used
+/// #[cfg(not(feature = "nothreads"))]
+/// #[gdextension(wasm_binary = "mycrate-with-threads.wasm")]
+/// unsafe impl ExtensionLibrary for MyExtension {}
+///
+/// // Binary name unchanged (mycrate.wasm) without thread support
+/// #[cfg(feature = "nothreads")]
+/// #[gdextension]
+/// unsafe impl ExtensionLibrary for MyExtension {}
+/// ```
````
I think the example is great and also readable, maybe mention briefly in a comment or the prose text, that `cfg_attr` could also be used as an alternative.

PgBiel replied:
(Not applicable anymore after the latest changes.)

Bromeon reviewed (commented):
Looks mostly good, thanks!

Bromeon commented:
Spontaneous thought, since `wasm_binary` is "dynamic" in nature (can be any string, not hardcoded identifier), did you consider making it a trait method of `ExtensionLibrary` rather than a proc-macro argument of `#[gdextension]`?

Or would that complicate fetching the value?

PgBiel commented:
> Spontaneous thought, since `wasm_binary` is "dynamic" in nature (can be any string, not hardcoded identifier), did you consider making it a trait method of `ExtensionLibrary` rather than a proc-macro argument of `#[gdextension]`?

Nope, did not. Just changed it and seems to work nicely. Thanks for the suggestion!

I considered making it a `const`, but it didn't seem consistent with the rest, so I made it a method.

PgBiel commented:
Been testing locally, worth mentioning. :)
